streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedTasks.java

```diff
@@ -421,28 +411,38 @@ void closeNonAssignedSuspendedTasks(final Map<TaskId, Set<TopicPartition>> newAs
                     suspendedTask.closeSuspended(true, false, null);
                 } catch (final Exception e) {
                     log.error("Failed to remove suspended {} {} due to the following error:", taskTypeName, suspendedTask.id(), e);
-                } finally {
-                    standByTaskIterator.remove();
                 }
+                standByTaskIterator.remove();
             }
         }
     }
 
     void close(final boolean clean) {
-        close(allTasks(), clean);
-        clear();
-    }
-
-    private void close(final Collection<Task> tasks, final boolean clean) {
-        for (final Task task : tasks) {
+        for (final Task task : allTasks()) {
             try {
                 task.close(clean, false);
+            } catch (final ProducerFencedException ignoreAndSwallow) {
+                closeZombieTask(task);
             } catch (final Throwable t) {
                 log.error("Failed while closing {} {} due to the following error:",
                           task.getClass().getSimpleName(),
                           task.id(),
                           t);
+                if (clean) {
```

Comment:
IMHO a little  easier to follow by putting `if(clean)...{ }` block in a private method, possible name `closeIfClean(clean, task, t)`